### PR TITLE
refactor: Add vacuum for volume material mapping

### LINKED ISSUE #431: Volume material mapping with vacuum
As I was was testing the volume mapping on the ODD I notice an issue when mapping material volume onto volume with vacuum. Since the vacuum is not consider a material by G4 no step are created when going through it and the it not part of the computation of the average material.

This can be solve by adding vacuum material hit between material hits if the material thickness of the first hit is smaller than the step size.

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
This PR resolve Issue #431 by adding additional vacuum material point between the already registered material slab if needed.

It also change the way material point are recorded. Instead of splitting the material hits into many different ones along the direction of the track, one material hit can now be associated with many different position. This greatly reduce the number of hits that need to be store for large volumes. 

Fixes #431.

### CODE REVIEW

**Comment by Corentin-Allaire** [comment]
If possible I would like if we could add this to the `for-corentin-no-cmake-changes` branch so I could test my changes to the volume mapping on ATLAS.

**Comment by msmk0** [comment]
> If possible I would like if we could add this to the `for-corentin-no-cmake-changes` branch so I could test my changes to the volume mapping on ATLAS.

Sure. I can either cherry-pick them into your branch or merge the current master into it (after this has been merge). With the later variant you also have all the additional changes that were added. Let me know what you prefer.

**Comment by Corentin-Allaire** [comment]
> > If possible I would like if we could add this to the `for-corentin-no-cmake-changes` branch so I could test my changes to the volume mapping on ATLAS.
> 
> Sure. I can either cherry-pick them into your branch or merge the current master into it (after this has been merge). With the later variant you also have all the additional changes that were added. Let me know what you prefer.

I would prefer if you could cherry-pick it now. That way I can test the new volume mapping on ATLAS before this get merged.

File: Core/include/Acts/Material/VolumeMaterialMapper.hpp
Comment (msmk0) [inline Core/include/Acts/Material/VolumeMaterialMapper.hpp:70]: Can you give this a more descriptive name, e.g. `BoundarySurfaceSelector` or something along those lines. Does this have to be a public type?

File: Core/include/Acts/Material/VolumeMaterialMapper.hpp
Comment (msmk0) [inline Core/include/Acts/Material/VolumeMaterialMapper.hpp:177]:
```suggestion
  /// Create extra material point for the mapping.
```
An explicit `@brief` is not required.

**Review by msmk0 — CHANGES_REQUESTED** [review]
This looks Ok to me. I can not do a thorough review since I do not know enough about the material mapping process.

@Corentin-Allaire You might want to request another review from somebody with more experience here. If you are confident that this is doing what it is supposed to do, we can also merge it in after fixing the minor comments.

File: Core/include/Acts/Material/VolumeMaterialMapper.hpp
Comment (Corentin-Allaire) [reply Core/include/Acts/Material/VolumeMaterialMapper.hpp:177]: Should I remove all the @brief from the volume mapping then ?

Comment (msmk0) [reply Core/include/Acts/Material/VolumeMaterialMapper.hpp:177]: In my view yes, but not in this PR. I just mentioned this here because it was changed.

File: Core/include/Acts/Material/VolumeMaterialMapper.hpp
Comment (Corentin-Allaire) [reply Core/include/Acts/Material/VolumeMaterialMapper.hpp:70]: No it does have to be public, I will switch it private with the material volume one too.

**Comment by Corentin-Allaire** [comment]
@msmk0 This should have resolve the conflict.

**Review by msmk0 — APPROVED** [review]

**Comment by asalzburger** [comment]
Updating the branch, but we should get this in next then.

**Pull request merged**
